src/main/kotlin/tanvd/grazi/ide/ui/GraziSettingsPanel.kt

```diff
@@ -1,31 +1,135 @@
 import com.intellij.openapi.options.ConfigurableUi
 import com.intellij.openapi.ui.ComboBox
 import com.intellij.ui.CheckBoxList
+import com.intellij.ui.ScrollPaneFactory
 import com.intellij.ui.components.JBCheckBox
 import com.intellij.ui.components.JBScrollPane
+import com.intellij.util.ui.JBUI
+import com.intellij.util.ui.UIUtil
+import kotlinx.html.*
+import org.apache.commons.text.similarity.LevenshteinDistance
+import org.languagetool.rules.Category
+import org.languagetool.rules.Rule
+import org.picocontainer.Disposable
 import tanvd.grazi.GraziConfig
 import tanvd.grazi.ide.GraziLifecycle
 import tanvd.grazi.language.Lang
-import java.awt.BorderLayout
-import java.awt.Font
-import javax.swing.BoxLayout
-import javax.swing.JComponent
+import tanvd.grazi.utils.*
+import java.awt.*
+import javax.swing.*
+import javax.swing.event.HyperlinkEvent
 
 
-class GraziSettingsPanel : ConfigurableUi<GraziConfig> {
+class GraziSettingsPanel : ConfigurableUi<GraziConfig>, Disposable {
     private val cbEnableGraziSpellcheck = JBCheckBox(msg("grazi.ui.settings.enable.text"))
     private val cmbNativeLanguage = ComboBox<Lang>()
     private val cblEnabledLanguages = CheckBoxList<String>()
 
+    private val descriptionPane = JEditorPane().apply {
+        editorKit = UIUtil.getHTMLEditorKit()
+        isEditable = false
+        isOpaque = false
+        border = JBUI.Borders.empty(UIUtil.DEFAULT_VGAP, UIUtil.DEFAULT_HGAP)
+    }.apply {
+        addHyperlinkListener { event ->
+            if (event?.eventType == HyperlinkEvent.EventType.ACTIVATED) {
+                Desktop.getDesktop()?.browse(event?.url?.toURI())
+            }
+        }
+    }
+
+    private val rulesTree = GraziRulesTree {
+        descriptionPane.text = when (it) {
+            is Rule -> html {
+                p {
+                    unsafe { +msg("grazi.ui.settings.rules.rule.template", it.description, it.category.name) }
+                }
+
+                it.url?.let {
+                    br
+                    p {
+                        +msg("grazi.ui.settings.rules.rule.description")
+                        +" "
+                        a(it.toString()) {
+                            unsafe { +it.toString() }
+                        }
+                    }
+                }
+
+                br
+
+                p {
+                    it.incorrectExamples?.let { examples ->
+                        if (examples.isNotEmpty()) {
+                            i {
+                                +msg("grazi.ui.settings.rules.rule.examples")
+                            }
+
+                            table {
+                                val accepted = ArrayList<String>()
+                                examples.forEach {
+                                    // remove very similar examples
+                                    if (!accepted.any { example -> LevenshteinDistance().apply(it.example, example).toDouble() / it.example.length < 0.2 }) {
```

Comment:
Should we construct LevensteinDistance() each time?

Reply:
Probably not